cognite_toolkit/_cdf_tk/utils/producer_worker.py

```diff
@@ -148,103 +167,144 @@ def user_input_listener() -> None:
-        if self.error_occurred:
+        if self._error_event.is_set():
             raise ToolkitRuntimeError(f"An error occurred during execution: {self.error_message}")
-        if self.stopped_by_user:
+        if self._stop_event.is_set():
             raise ToolkitRuntimeError("Execution was stopped by the user.")
 
     def _download_worker(self, progress: Progress, download_task: TaskID) -> None:
         """Worker thread for downloading data."""
         try:
             iterator = iter(self._download_iterable)
         except TypeError as e:
-            self.error_occurred = True
+            self._error_event.set()
             self.error_message = str(e)
             self.console.print(f"[red]Error[/red] occurred while {self.download_description}: {self.error_message}")
             return
         item_count = 0
-        while not self.error_occurred and not self._stop_event.is_set():
+        while not self._error_event.is_set():
             try:
+                if self._stop_event.is_set():
+                    break
                 items = next(iterator)
                 self.total_items += len(items)
-                item_count += len(items)
-                while not self.error_occurred:
-                    try:
-                        self.process_queue.put(items, timeout=0.5)
-                        progress.update(download_task, advance=1, item_count=item_count)
-                        break  # Exit the loop once the item is successfully added
-                    except queue.Full:
-                        # Retry until the queue has space
-                        continue
+                if self._put_with_error_check(items, self.process_queue):
+                    item_count += len(items)
+                    progress.update(download_task, advance=1, item_count=item_count)
+                    continue
+                break  # Exit if error event was set while waiting to put
             except StopIteration:
-                self.download_terminated = True
                 break
             except Exception as e:
-                self.error_occurred = True
+                self._error_event.set()
                 self.error_message = str(e)
                 self.console.print(f"[red]Error[/red] occurred while {self.download_description}: {self.error_message}")
                 break
-        if self._stop_event.is_set():
-            self.download_terminated = True
+        self._put_with_error_check(PROCESS_FINISH_SENTINEL, self.process_queue)  # type: ignore[misc]
+
+    def _put_with_error_check(self, items: T_Item, target_queue: queue.Queue[T_Item]) -> bool:
+        """Helper to put items into a queue with error checking."""
+        while not self._error_event.is_set():
+            try:
+                target_queue.put(items, timeout=0.5)
+                return True
+            except queue.Full:
+                continue
+        return False
```

Comment:
Instead of using `queue.put(items)` which is a blocking call, this ensure that we check if any error has occurred. The previous version of WorkerProducerExecutor could freeze if an error occurred while whiting to put an item on the queue.